File: 2022/python/day8.py
with open('../inputs/input8') as f:
	input = f.readlines()
	forest = [[int(t) for t in line.strip()] for line in input if len(line) > 0]
	visible = [[False for _ in row] for row in forest]

	for y in range(len(forest)):
		max_left = -1
		max_right = -1
		for x in range(len(forest[y])):
			for xl in range(0, x):
				if forest[y][xl] > max_left:
					max_left = forest[y][xl]
					visible[y][xl] = True
			for xr in range(len(forest[y])-1, x, -1):
				if forest[y][xr] > max_right:
					max_right = forest[y][xr]
					visible[y][xr] = True

	for x in range(len(forest[0])):
		max_top= -1
		max_bottom = -1
		for y in range(len(forest)):
			for yu in range(0, y):
				if forest[yu][x] > max_top:
					max_top = forest[yu][x]
					visible[yu][x] = True
			for yb in range(len(forest)-1, y, -1):
				if forest[yb][x] > max_bottom:
					max_bottom = forest[yb][x]
					visible[yb][x] = True					

	print('Part 1:', sum([col for row in visible for col in row if col]))

	scenic_score = 0
	for y in range(len(forest)):
		for x in range(len(forest[y])):

			up, down, left, right = 0,0,0,0

			for yv in range(y-1, -1, -1):
				if forest[yv][x] < forest[y][x]:
					up+=1
				else:
					up+=1
					break
			for yv in range(y+1, len(forest)):
				if forest[yv][x] < forest[y][x]:
					down+=1
				else:
					down+=1
					break
			for xh in range(x-1, -1, -1):
				if forest[y][xh] < forest[y][x]:
					left+=1
				else:
					left+=1
					break
			for xh in range(x+1, len(forest[y])):
				if forest[y][xh] < forest[y][x]:
					right+=1
				else:
					right+=1
					break
			scenic_score = max(scenic_score, up * down * left * right)

	print('Part 2:', scenic_score)

File: 2022/ts/day8.ts
import { assert } from "./deps.ts";
import { readLines } from "./utils.ts";

export function run(lines: string[]): number {
  let result = 0;

  const treeGrid = lines.map((line) =>
    line.split("").map((token) => parseInt(token))
  );

  const visibleGrid: boolean[][] = [];
  for (let y = 0; y < treeGrid.length; ++y) {
    const row = [];
    for (let x = 0; x < treeGrid[y].length; ++x) {
      row.push(false);
    }
    visibleGrid.push(row);
  }

  const mark = (x: number, y: number, max: number): number => {
    const tree = treeGrid[y][x];
    const visible = visibleGrid[y][x];
    if (tree > max) {
      max = tree;
      if (!visible) {
        visibleGrid[y][x] = true;
      }
    }
    return max;
  };

  for (let y = 0; y < treeGrid.length; ++y) {
    let currMax = -1;
    for (let x = 0; x < treeGrid[y].length; ++x) {
      currMax = mark(x, y, currMax);
    }
    currMax = -1;
    for (let x = treeGrid[y].length - 1; x >= 0; --x) {
      currMax = mark(x, y, currMax);
    }
  }

  for (let x = 0; x < treeGrid[0].length; ++x) {
    let currMax = -1;
    for (let y = 0; y < treeGrid.length; ++y) {
      currMax = mark(x, y, currMax);
    }
    currMax = -1;
    for (let y = treeGrid.length - 1; y >= 0; --y) {
      currMax = mark(x, y, currMax);
    }
  }

  for (let y = 0; y < visibleGrid.length; ++y) {
    for (let x = 0; x < visibleGrid[y].length; ++x) {
      if (visibleGrid[y][x]) {
        result++;
      }
    }
  }

  return result;
}

function part1() {
  const part1 = run(
    readLines("../inputs/input8").filter((line) => line.length > 0),
  );
  console.log(`Part 1: ${part1}`);
}

part1();

export function run2(lines: string[]): number {
  let result = 0;

  const treeGrid = lines.map((line) =>
    line.split("").map((token) => parseInt(token))
  );

  for (let y = 0; y < treeGrid.length; ++y) {
    for (let x = 0; x < treeGrid[y].length; ++x) {
      const tree = treeGrid[y][x];

      let countUp = 0;
      for (let ys = y - 1; ys >= 0; --ys) {
        const currTree = treeGrid[ys][x];
        if (currTree < tree) {
          countUp++;
        } else {
          countUp++;
          break;
        }
      }

      let countBottom = 0;
      for (let ys = y + 1; ys < treeGrid.length; ++ys) {
        const currTree = treeGrid[ys][x];
        if (currTree < tree) {
          countBottom++;
        } else {
          countBottom++;
          break;
        }
      }

      let countLeft = 0;
      for (let xs = x - 1; xs >= 0; --xs) {
        const currTree = treeGrid[y][xs];
        if (currTree < tree) {
          countLeft++;
        } else {
          countLeft++;
          break;
        }
      }

      let countRight = 0;
      for (let xs = x + 1; xs < treeGrid[y].length; ++xs) {
        const currTree = treeGrid[y][xs];
        if (currTree < tree) {
          countRight++;
        } else {
          countRight++;
          break;
        }
      }
      const score = countUp * countLeft * countRight * countBottom;
      result = Math.max(result, score);
    }
  }

  return result;
}

function part2() {
  const part2 = run2(
    readLines("../inputs/input8").filter((line) => line.length > 0),
  );
  console.log(`Part 2: ${part2}`);
}

part2();
